pkg/controllers/provisioning/scheduling/nodeset.go

```diff
@@ -76,3 +102,583 @@ func (s *NodeSet) getDaemons(ctx context.Context, constraints *v1alpha5.Constrai
 func (s *NodeSet) Add(node *Node) {
 	s.nodes = append(s.nodes, node)
 }
+
+func (s *NodeSet) SchedulePod(ctx context.Context, pod *v1.Pod) error {
+	// copy the pod as this method modifies the pod's node selectors to force it to meet topology constraints
+	pod = pod.DeepCopy()
+
+	var possibleNodes []*Node
+	for _, node := range s.nodes {
+		if err := node.Compatible(ctx, pod); err == nil {
+			possibleNodes = append(possibleNodes, node)
+		}
+	}
+
+	tightened := s.constraints.Requirements.Add(v1alpha5.NewPodRequirements(pod).Requirements...)
+	var err error
+	possibleNodes, err = s.filterByTopologies(ctx, tightened, pod, possibleNodes)
+	if err != nil {
+		return err
+	}
+
+	if len(possibleNodes) > 0 {
+		sort.Slice(possibleNodes, byPreferredNode(possibleNodes))
+		if err := possibleNodes[0].Add(ctx, pod); err != nil {
+			return fmt.Errorf("adding pod to node, %w", err)
+		}
+	} else {
+		n := NewNode(s.kubeClient, s.constraints, s.daemonResources, s.instanceTypes)
+		if err := n.Add(ctx, pod); err != nil {
+			return fmt.Errorf("adding pod to node, %w", err)
+		}
+		s.nodes = append(s.nodes, n)
+	}
+
+	if err := s.recordTopologyDecisions(pod); err != nil {
+		s.logger.Errorf("recording topology decision, %s", err)
+	}
+
+	return nil
+}
+
+func (s *NodeSet) TrackTopologies(ctx context.Context, pods []*v1.Pod) error {
+	var errs error
+	errs = multierr.Append(errs, s.trackExistingAntiAffinities(ctx))
+	for _, p := range pods {
+		// need to first ensure that we know all of the topology constraints.  This may require looking up
+		// existing pods in the running cluster to determine zonal topology skew.
+		if err := s.trackTopologySpread(ctx, p); err != nil {
+			errs = multierr.Append(errs, fmt.Errorf("tracking topology spread, %w", err))
+		}
+
+		if err := s.trackPodAffinityTopology(ctx, p); err != nil {
+			errs = multierr.Append(errs, fmt.Errorf("tracking affinity topology, %w", err))
+		}
+	}
+	return errs
+}
+
+// trackExistingAntiAffinities is used to identify pods with anti-affinity terms so we can track those topologies.  We
+// have to look at every pod in the cluster as there is no way to query for a pod with anti-affinity terms.
+func (s *NodeSet) trackExistingAntiAffinities(ctx context.Context) error {
+	// TODO: if can we determine somehow that the LimitPodHardAntiAffinityTopology admission controller is in place, we
+	// don't need to do any of this.  In that case anti-affinity would be hostname only which makes this tracking un-needed.
+	var nodeList v1.NodeList
```

Comment:
Thoughts on separating out topology into a state object to help us avoid calling List nodes so many times?

Reply:
It would be complicated.  We need to know the current state of the cluster with respect to all pods that have been scheduled by us or kube-scheduler previously that have pod anti-affinity.  We would have to watch for node creation/deletion, and pod creation/update/deletion.